ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,476 @@
+    new Random().nextBytes(userMetadata);
+    new Random().nextBytes(data);
+    BlobProperties properties = new BlobProperties(blobSize, "serviceid1");
+    OperationInfo opInfo = new OperationInfo(operationId, properties, userMetadata, data, opChain);
+    continueChain(opInfo);
+    return opInfo;
+  }
+
+  static Properties getRouterProperties(String routerDatacenter) {
+    Properties properties = new Properties();
+    properties.setProperty("router.hostname", "localhost");
+    properties.setProperty("router.datacenter.name", routerDatacenter);
+    properties.setProperty("router.connection.checkout.timeout.ms", "5000");
+    properties.setProperty("router.request.timeout.ms", "20000");
+    properties.setProperty("router.max.put.chunk.size.bytes", Integer.toString(CHUNK_SIZE));
+    properties.setProperty("router.put.success.target", "1");
+    properties.setProperty("coordinator.hostname", "localhost");
+    properties.setProperty("coordinator.datacenter.name", routerDatacenter);
+    return properties;
+  }
+
+  private static void checkBlobId(String blobId, String operationName) {
+    Assert.assertNotNull("Null blobId for operation: " + operationName, blobId);
+  }
+
+  private static void checkBlobInfo(BlobInfo blobInfo, OperationInfo operationInfo, String operationName) {
+    Assert.assertNotNull("Null blobInfo for operation: " + operationName, blobInfo);
+    Assert.assertEquals("Blob size in info does not match expected for operation: " + operationName,
+        operationInfo.properties.getBlobSize(), blobInfo.getBlobProperties().getBlobSize());
+    Assert.assertEquals("Service ID in info does not match expected for operation: " + operationName,
+        operationInfo.properties.getServiceId(), blobInfo.getBlobProperties().getServiceId());
+    Assert.assertArrayEquals("Unexpected user metadata for operation: " + operationName, operationInfo.userMetadata,
+        blobInfo.getUserMetadata());
+  }
+
+  private static void checkBlob(ReadableStreamChannel channel, OperationInfo operationInfo, String operationName) {
+    Assert.assertNotNull("Null channel for operation: " + operationName, channel);
+    try {
+      ByteBufferAsyncWritableChannel getChannel = new ByteBufferAsyncWritableChannel();
+      Future<Long> readIntoFuture = channel.readInto(getChannel, null);
+      int readBytes = 0;
+      do {
+        ByteBuffer buf = getChannel.getNextChunk();
+        int bufLength = buf.remaining();
+        Assert.assertTrue(
+            "total content read should not be greater than length of put content, operation: " + operationName,
+            readBytes + bufLength <= operationInfo.data.length);
+        while (buf.hasRemaining()) {
+          Assert.assertEquals("Get and Put blob content should match, operation: " + operationName,
+              operationInfo.data[readBytes++], buf.get());
+        }
+        getChannel.resolveOldestChunk(null);
+      } while (readBytes < operationInfo.data.length);
```

Comment:
For any misbehavior if a router returns bytes less than expected, would this loop be running forever?